README.md

```diff
@@ -1,30 +1,205 @@
-Definitely want to make use of a controlled component or two (which'll require
-state.)
+Instead of directly modifying the state using `this.state`, we use `this.setState()`. This is a function available to
+all React components, and allows us to let React know that the component state has changed. This way the components
+knows it should re-render, because its state has changed and its UI will most likely also change. Using a setter
+function like this is very performant: while other frameworks use dirty checking to see if a property has changed, React
+_already knows_ because we used a built-in function to let it know what changes we'd like to make!
 
-## Resources
+For example, let's say we have a component with a button, and a bit of text to indicate whether that button has been
+pressed yet:
+
+```js
+class ClickityClick extends React.Component {
+  constructor() {
+    super();
+    
+    // Define the initial state:
+    this.state = {
+      hasBeenClicked: false,
+    };
+    
+    this.handleClick = this.handleClick.bind(this);
+  }
+  
+  handleClick() {
+    // Update our state here...
+  }
+
+  render() {
+    return (
+      <div>
+        <p>I have {this.state.hasBeenClicked ? 'not' : null} been clicked yet!</p>
+        <button onClick={this.handleClick}>Click me!</button>
+      </div>
+    );
+  }
+}
+```
+
+To update our state, we use `this.setState()` and pass in an object. This object will get merged with the current state.
+When the state has been updated, our component re-renders automatically. Handy!
+
+```js
+handleClick() {
+  this.setState({
+    hasBeenClicked: true,
+  });
+}
+```
+
+## How state gets merged
+When updating state, we don't have to pass in the entire state, just the property we want to update. For example,
+consider the following state for our component:
+
+```
+{
+  hasBeenClicked: false,
+  currentTheme: 'blue',
+}
+```
+
+If we updated the `hasBeenClicked` using `this.setState()` like we did above, it would _merge_ the new state with the
+existing state, resulting in this new state:
+
+```
+{
+  hasBeenClicked: true,
+  currentTheme: 'blue',
+}
+```
+
+One super important thing to note is that it only merges things on the first level. Let's say we're working on a
+component that lets a user fill in an address, and the component's state is structured like this:
+
+```
+{
+  theme: 'blue',
+  addressInfo: {
+    street: null,
+    number: null,
+    city: null,
+    country: null
+  },
+}
+```
+
+If we wanted to update the `addressInfo.city` field, you might think we can update it like this:
+
+```js
+this.setState({
+  address: {
+    city: 'New York City',
+  },
+});
+```
+
+**However**, this would result in the following state shape:
+
+```
+{
+  theme: 'blue',
+  addressInfo: {
+    city: 'New York City',
+  },
+}
+```
 
+See what happened there? It _merged_ the state, but any objects get overwritten, because it doesn't _deeply_ merge the
```

Comment:
Will students understand what deep merge is? Is this an opportunity to link out to another resource that might give them an idea? (Now that I think of it, maybe you could do the same for the "dirty checking" concept above?)